book/src/proposals/validator-timestamp-oracle.md

```diff
@@ -0,0 +1,110 @@
+- Each validator must maintain a timestamp oracle
+
+## Recording Time
+
+At regular intervals, each validator records its observed time by submitting a
+Timestamp transaction to the cluster as it is processing a particular slot. The
+transaction contains the slot index and the validator's system time as Unix
+timestamp (seconds since the Unix epoch), and is signed by the validator's
+identity keypair.
+
+This proposal suggests that Timestamp transactions be issued every 30min, which
+should be short enough to prevent block times drifting very much, without adding
+too much transaction overhead to the cluster. Validators can convert this time
+to a slot interval using the `slots_per_year` value that is stored in each bank.
+
+```text
+let seconds_in_30min = 1800;
+let timestamp_interval = (slots_per_year / SECONDS_PER_YEAR) * seconds_in_30min;
+```
+
+This transaction should be triggered in replay_stage (likely near where
+`subscriptions.notify_slot()` is executed) when `current_slot %
+timestamp_interval == 0`.
+
+### Timestamp Accounts
+
+Before beginning operation, a validator needs to create and initialize a
+Timestamp account that will hold its most recent slot-timestamp. Like the
+validator's vote account, this account needs to be rent exempt.
+
+### Timestamp Program
+
+The on-chain Timestamp program processes Timestamp transactions from validators.
+It needs to load the correct Timestamp account, verify that the transaction
+signer is the expected validator, compare the slot and timestamp to the
+currently stored values to verify that they are both monotonically increasing,
+and store the new slot and timestamp in the account.
+
+## Calculating Median Timestamp
+
+In order to calculate the estimated timestamp for a particular block, a
+validator first needs to identify the most recently timestamped slot:
+
+```text
+let timestamp_slot = floor(current_slot / timestamp_interval);
+```
+
+Then the validator needs to gather all Timestamp transactions from the ledger
+that reference that slot, using `Blocktree::get_slot_entries()`. As these
+transactions could have taken some time to reach and be processed by the leader,
+the validator needs to scan several completed blocks after the timestamp_slot to
+get a reasonable set of Timestamps. The exact number of slots will need to be
+tuned: More slots will enable greater cluster participation and more timestamp
+datapoints; fewer slots will speed how long Timestamp filtering takes.
+
+From this collection of transactions, the validator selects the median timestamp.
+
+Any validator replaying the ledger should derive the same median timestamp by processing the Timestamp transactions from the same number of slots.
+
+## Calculating Estimated Time for a Particular Block
+
+Once the median timestamp for a known slot is calculated, it is trivial to
+calculate the estimated timestamp for subsequent block N:
+
+```text
+let block_n_timestamp = median_timestamp + (block_n_slot_offset * slot_duration);
```

Comment:
Might be useful to calculate an individual validator's timestamp first, and then find the average. That'd allow validators to submit timestamps on arbitrary slots, instead of requiring them to stamp the same slot.

Reply:
The main benefit of doing it the way it's written (established timestamp slots) is that it's easier to target likely slots to scan for Timestamp transactions. I suppose another alternative would be to start at the block whose time is desired and scan backwards until we get some baseline number of timestamps, but this could result in having to scan every slot in the expected timestamp interval.